pkg/api/types.go

```diff
@@ -180,6 +180,112 @@ type VolumeSource struct {
 	GitRepo *GitRepo `json:"gitRepo"`
 	// Secret represents a secret that should populate this volume.
 	Secret *SecretSource `json:"secret"`
+
+	// TODO (per thockin) https://github.com/GoogleCloudPlatform/kubernetes/pull/4055#discussion_r24865963
+	AWSElasticBlockStore *AWSElasticBlockStore `json:"elasticBlockStore"`
+
+	NFSMount *NFSMount `json:"nfsMount"`
+
+	PersistentVolumeClaimAttachment *PersistentVolumeClaimAttachment `json:"persistentVolumeClaim,omitempty"`
+}
+
+type PersistentVolume struct {
+	TypeMeta   `json:",inline"`
+	ObjectMeta `json:"metadata,omitempty"`
+
+	//Spec defines a persistent volume owned by the cluster
+	Spec PersistentVolumeSpec `json:"spec,omitempty"`
+
+	// Status represents the current information about persistent volume.
+	Status PersistentVolumeStatus `json:"status,omitempty"`
+}
+
+type PersistentVolumeSpec struct {
+	Capacity ResourceList `json:"capacity,omitempty`
+	// AccessModeTypes are inferred from the Source
+	Source VolumeSource `json:",inline"`
```

Comment:
Thanks for the explanation! I understand you want to make the access modes provider specific. I still have one lingering concern about VolumeSource being overloaded. Lets take GCE as a possible PV provider.

If a Persistent Volume is created using these APIs with GCE as provider, here is how the PodSpec will look:
{
   "volumes": [{
         "source": {
               "persistentVolumeClaim": {
                     "spec": {
                          ....
                      }
                  }
            }
      }]
}

If the same persistent volume was creating outside of k8s (may be using GCE APIs), the Pod spec will look like it does today.
{
   "volumes": [{
         "source": {
             "persistentDisk":  {
                  ...
              }
        }]
}

Now lets assume there is storage administrator who is responsible for provisioning storage. Does the Pod user need to know whether the volume was provisioned with k8s or GCE and use an appropriate PodSpec?

One way to simplify it is to make VolumeSource only refer to temporary/ephemeral storage requirements or a persistent volume attachment. PersistentVolumeSpec requirements could come from another structure, may be PersistentVolumeSource?